prusti-viper/src/encoder/encoder.rs

```diff
@@ -252,10 +263,144 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
                 type_vars: vec![],
             });
         }
+
+        if config::enable_purification_optimization() {
+            domains.push(self.axiomatized_function_domain.borrow().clone());
+            let builtin_encoder =  BuiltinEncoder::new();
+            domains.push(builtin_encoder.encode_builtin_domain(BuiltinDomainKind::Nat));
+            domains.push(builtin_encoder.encode_builtin_domain(BuiltinDomainKind::Primitive));
+        }
+
         domains.sort_by_key(|d| d.get_identifier());
         domains
     }
 
+    pub fn encode_axiomatized_pure_function(&self, f: &vir::Function) {
+        let snapshots: &HashMap<String, Box<Snapshot>> = &self.snapshots.borrow();
+        let domain_name = self.axiomatized_function_domain.borrow().name.clone();
+
+        let formal_args_without_nat: Vec<vir::LocalVar> =
+            snapshot::encode_axiomatized_function_args_without_nat(&f.formal_args, &snapshots);
+
+        let df =
+            snapshot::encode_axiomatized_function(&f.name, &f.formal_args, &f.return_type, &snapshots);
+        let nat_arg = vir::Expr::local(snapshot::encode_nat_argument());
+        let nat_succ = vir::Expr::domain_func_app(snapshot::get_succ_func(), vec![nat_arg.clone()]);
+        let args_without_nat: Vec<vir::Expr> = formal_args_without_nat
+            .clone()
+            .into_iter()
+            .map(vir::Expr::local)
+            .collect();
+
+        let mut args_with_succ = args_without_nat.clone();
+        args_with_succ.push(nat_succ);
+
+        let function_call_with_succ = vir::Expr::domain_func_app(df.clone(), args_with_succ.clone());
+
+        let mut args_with_zero = args_without_nat.clone();
+        args_with_zero.push(vir::Expr::domain_func_app(snapshot::get_zero_func(), Vec::new()));
+
+        let function_call_with_zero = vir::Expr::domain_func_app(df.clone(), args_with_zero.clone());
+
+        let mut purifier = snapshot::ExprPurifier {
+            snapshots: &snapshots,
+            self_function: function_call_with_succ.clone(),
+        };
+
+        let pre_conds: vir::Expr = f
+            .pres
+            .iter()
+            .cloned()
+            .map(|p| vir::ExprFolder::fold(&mut purifier, p))
+            .conjoin();
+        let post_conds: vir::Expr = f
+            .posts
+            .iter()
+            .cloned()
+            .enumerate()
+            .filter_map(|(i, p)| {
+                if i == f.posts.len() - 1 {
+                    // Skip the last post condition as it is only there to clarify the relation between the result of this function and snapshots
+                    None
+                } else {
+                    Some(vir::ExprFolder::fold(&mut purifier, p))
+                }
+            })
+            .conjoin();
+
+        let function_body = vir::ExprFolder::fold(&mut purifier, f.body.clone().unwrap());
+
+        let function_identiry = vir::Expr::eq_cmp(function_call_with_succ.clone(), function_body);
```

Comment:
```suggestion
        let function_identity = vir::Expr::eq_cmp(function_call_with_succ.clone(), function_body);
```

Also (incl. this one) many cases of unnecessary variables that are just used once.